### ISSUE TITLE
Increase the payload?

### ISSUE THREAD

**Issue by sphh** [electrotype/thunderbird-addon-scriptable-notifications#4]
The information currently sent to the script is good enough for a simple script.

I have a script, which could display the number of messages, create notifications etc. For this it would need more information from ScriptableNotifications. At the moment it cannot even tally and display the number of unread messages, because the script receives a `true` even for read messages.

Do you think, it would be possible to increase the payload with information about the new/read/deleted message? Obviously this should be an opt-in feature, so that current scripts do not break.

I could think of the following interesting information:
- Message ID
- Sender
- To, Cc or Bcc?
- Subject
- Priority
- Status (new|read|deleted)
- Mailbox
and maybe some more

**Comment by electrotype** [comment]
Hello,

I do not plan on implementing such kind of features for now, sadly. I'm quite busy on another personnal project. But I'm open to Pull Requests, of course! Same thing for your other issue (listing all folders - probably not the "trash" ones though): if you are able to list them in a nice way (in a tree?), I would look at a PR for sure.

Note that passing complex information to a _batch file_ is not trivial (look at https://github.com/electrotype/thunderbird-addon-scriptable-notifications/blob/main/scriptExamples/linux/simpleMsg/script.sh#L43-L46 and https://stackoverflow.com/a/24777120/843699 ) and I DO want the add-on to support calls to a regular batch file (not just Python files and more complex languages). It is an interesting challenge though, if you want to try.

I may have a look to this request again in the future, if I have more time.

Cheers!

**Comment by sphh** [comment]
I am with you, that retrieving complex information from a _batch file_ is not trivial. So the current behaviour should stay the default and an optional interface for powerful programs could be added. This could be activated in the configuration screen.

Unfortunately my knowledge of JavaScript is limited and my knowledge of Thunderbird's internals are even more limited, so it would certainly be a challenge (= it will be way above my capabilities).

If my Python script proves to work reliably – and I find the time to clean it up and document it –, I might make a PR with it as a prove of concept (and a teaser to show, what might be possible with a more complex language).

**Comment by electrotype** [comment]
I agree an option to decide between simple and complex data to send (leading to support or not for batch files) could be an idea. But finding a way to parse more complex data in batch files, and staying compatible with them, would be the ideal imo. Maybe a simple "key=value\nkey=value\nkey=value" data stream with a special character at the end so the script can detect the end of stdin data... I don't know.

Feel free to make a PR indeed, even if it's only for a script example that _could_ be supported one day.

**Comment by sphh** [comment]
Would it work with simple batch scripts, if the JSON object were:
```json
"true" : {<complex data>}
```
or
```json
"false" : {<complex data>}
```
with `true` and `false` as it is now. The simple batch script looks only at the first value, discarding the rest. If that works, there would be no need to switch between interfaces.

If a batch/shell script wants to access the JSON data, it could parse it with [jq](https://stedolan.github.io/jq/) (or any other tool listed – or not listed – [here](https://medium.com/swlh/different-ways-to-handle-json-in-a-linux-shell-81183bc2c9bc)).

I had a look at the code of ScriptableNotifications. I could add an option `simple/complex` to the option screen and prepare a JSON object with the structure of the complex data – or just prepare the JSON object, if this option is not an option. Most likely it would need your input to actually fill it with the data.

**Comment by electrotype** [comment]
Lol, I like how you think! That's pretty creative/hacking.

But I think adding an option for "simple/complex" data would be cleaner, if parsing complex data is too hard in a batch file, after all.

**Comment by sphh** [comment]
I would call it hackish. :stuck_out_tongue_winking_eye:

I also believe, that an option is much, much cleaner!

**Comment by electrotype** [comment]
*hackish yes. I'm still working on my english!

**Comment by electrotype** [comment]
By the way, I wrote "batch" file in the previous messages while I wanted to say "bash". I'll correct those.

**Comment by sphh** [comment]
Your English is perfect. IMHO there is only a slightly semantic difference and ‘hackish’ is a little bit more dodgy compared to ‘hacking’.

**Comment by sphh** [comment]
I was always thinking `bash` anyway, but I believe the same limitation (if not more) can be found in batch files!

**Comment by sphh** [comment]
I managed to implement the "extended" feature. Please see https://github.com/sphh/thunderbird-addon-scriptable-notifications/tree/extended.

There is one big problem: The listener `newEmailListener` (`browser.messages.onNewMailReceived.addListener()`) receives a list with _all_ new messages and not just the recent new message. This is a somewhat unexpected behaviour, as also encountered by you and others (https://stackoverflow.com/questions/65090864/onnewmailreceived-event-returns-message-list-containing-messages-from-previous). For sending information about the new message, knowing the new message is essential. Do you have a clue, how to get hold of it?

My only idea was to make a list (or better set) of mail IDs of all new messages for each watched folder at start up; a so-called "seenMails" list. A comparison of this stored list with the list of messages sent to `newEmailListener` should reveal the new message (hopefully just one). The "seenMails" set then updated and the new message passed on to the external script.

A bit awkward … Do you have a better idea?

Because of this hiccup, I have disabled the `newEmailListener` for the moment.

Some other notes:
- There are lines marked with `// BUG: `. Do you have any comments on them?
- The test messages are not coded yet.

**Comment by sphh** [comment]
The information sent to the external script is the following JSON string:
```json
{"accounts": [
    {"id": "account1",
     "identities": [
        {"email": "name.surname@company.com",
         "label": "",
         "name": "Name Surname",
         "organization": "Company"}],
     "name": "Business",
     "type": "imap"},
    {"id": "account3",
     "identities": [
        {"email": "name@private.net",
         "label": "",
         "name": "Name Surname",
         "organization": ""},
        {"email": "surname@private.net",
         "label": "",
         "name": "Name Surname",
         "organization": ""},
        {"email": "name.surname@private.net",
         "label": "",
         "name": "Name Surname",
         "organization": ""}],
     "name": "Private",
     "type": "imap"},
    {"id": "account2",
     "identities": [],
     "name": "Local Folders",
     "type": "none"}],
 "folders": [
    {"accountId": "account1",
     "favorite":true,
     "name": "Inbox",
     "path": "/INBOX",
     "totalMessageCount": 74,
     "type": "inbox",
     "unreadMessageCount": 7},
    {"accountId": "account3",
     "favorite":true,
     "name": "Inbox",
     "path": "/INBOX",
     "totalMessageCount": 94,
     "type": "inbox",
     "unreadMessageCount": 3}],
 "event": "read",
 "message": {
    "author": "Someone Else <some.one.else@nowhere.org>",
    "ccList": [],
    "date": "2022-10-12T16:05:00.000Z",
    "flagged": false,
    "messageId": "9477b273-0cea-c454-e6c3-86f452807092@nowhere.org",
    "headersOnly": false,
    "junk": false,
    "junkScore": 0,
    "read": true,
    "size": 4014,
    "subject": "ScriN #9",
    "tags": [],
    "folder": {
        "accountId": "account3",
        "name": "Inbox",
        "path": "/INBOX",
        "type": "inbox"}
    }
}
```

Notes:
- The value of the key `event` is `"start"`, `"new"`, `"read"` or `"deleted"`.
- The value of `message` is `null` for the `"start"` event.

PS: I also wonder, if we have to watch the `onMoved` event? (Or is a `onDeleted` event fired, if a message is moved away?)

**Comment by electrotype** [comment]
For some reason I didn't get any email for your last replies. I'll answer soon!

**Comment by electrotype** [comment]
It would be easier to comment on a proper PR, but here are my comments for the modifications on your "extended" branch.

1. I won't have time to work on scripts that would use this new "extended" type, so you would have to provide at least a simple one. A Python script that displays an alert containing the entire JSON as received would be perfect!

2. You would have to update the documentation too.

3. Rename/change some name/comments if requried. For example "saveFolders()" should now be "saveOptions()" I guess.

4. Indeed "newEmailListener" is crucial, you would have to be enabled back. Your "seenMails" store would work I guess, but it would have to be cleaned up (on message read/deleted?). It should also not be used by the "simple" type.

5. When Thunderbird starts, are you sure you do not want to call the script if there are still unread "new" messages in the inboxes? I understand the script has already been called for those, but I don't know... Maybe the strating event  should be a special case? Anyway, you know what you need... I will still be using the "simple" type so be free to decide what this first version of the "extended" version does or not.

6. In the UI, it is not obvious at all that the user must click the "Save" button in the following section in order to save the "type" option! I would automatically save the option when it changes. Maybe we should do that for the folders too then.

7. I agree that the code to change the "disabled" attributes of some buttons may be cleaner than that. But that wouldn't stop a PR from being accepted.

So I encourage you to continue with those ideas and make a proper PR (that we could still tweak before merging).

Thank you!

**Comment by sphh** [comment]
Thanks for your comments.

>     1. I won't have time to work on scripts that would use this new "extended" type, so you would have to provide at least a simple one. A Python script that displays an alert containing the entire JSON as received would be perfect!

That should be possible.

>     2. You would have to update the documentation too.

I'll have a look.

>     3. Rename/change some name/comments if required. For example "saveFolders()" should now be "saveOptions()" I guess.

I left them as original on purpose, but I am happy to change these.

>     4. Indeed "newEmailListener" is crucial, you would have to enable it back. Your "seenMails" store would work I guess, but it would have to be cleaned up (on message read/deleted?). It should also not be used by the "simple" type.

So I guess, you do not have any other (better) idea. :frowning:

>     5. When Thunderbird starts, are you sure you do not want to call the script if there are still unread "new" messages in the inboxes? I understand the script has already been called for those, but I don't know... Maybe the start event should be a special case? Anyway, you know what you need... I will still be using the "simple" type so feel free to decide what this first version of the "extended" version does or not.

Do you mean to call the script with each of the unread messages? That was my first idea, but then I noticed, that there could be thousands of unread messages, so I decided to just pass the number of total and the number of unread messages. I believe, that is a sensible decision, because if you start Thunderbird, you are just interested, if there are unread messages. I am happy to pass every unread message, but fear, that this would slow down Thunderbird.

What do you think?

>     6. In the UI, it is not obvious at all that the user must click the "Save" button in the following section in order to save the "type" option! I would automatically save the option when it changes. Maybe we should do that for the folders too then.

Mmh, I thought, if there is a "Save" button, it is obvious. Obviously not.

I would suggest the following approach:
1. Split out steps 1. and 2.. These are only needed, when started and then only, if someone wants to reconfigure them. A section, which is hidden would be suitable (and show them, if no script was generated before).
2. Put step 3. under a separate heading (e.g. "Configure native script").
3. If the sections for configuring the native script and to select the mailboxes and the "Save" button are - let's say - within a frame, it should be obvious.

I do not object to save the option and the folders when any of them changes. But I'm afraid, it will not be something I will implement. Sorry.

>     7. I agree that the code to change the "disabled" attributes of some buttons may be cleaner than that. But that wouldn't stop a PR from being accepted.

Any ideas how to do that?

> So I encourage you to continue with those ideas and make a proper PR (that we could still tweak before merging).

Ok, I'll make a PR with my changes.

**Comment by electrotype** [comment]
Extended mode now available in version 4.0.

Thanks to @sphh for this feature!
